src/vivarium/framework/components/manager.py

```diff
@@ -81,26 +81,34 @@ def get_components(self, component_type: str):
         return self._component_manager.get_components(component_type)
 
 
-def _setup_components(builder, component_list, configuration):
-    done = []
-    while component_list:
-        component = component_list.pop(0)
-        if component is None:
+def _setup_components(builder, component_list: Sequence, configuration: ConfigTree) -> Sequence:
+
+    configured = []
+    for c in component_list:  # apply top-level configurations first
+        if hasattr(c, "configuration_defaults") and not c in configured:
+            _apply_component_default_configuration(configuration, c)
+            configured.append(c)
+
+    setup = []
+    while component_list:  # mutated at runtime by calls to setup
```

Comment:
thanks for adding this comment! was confused why you had to keep track of what had already been setup